ledger/participant-state/kvutils/tools/integrity-check/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/LogAppendingReadServiceFactory.scala

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.tools.integritycheck
+
+import akka.NotUsed
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import com.daml.ledger.api.health.HealthStatus
+import com.daml.ledger.participant.state.kvutils.OffsetBuilder
+import com.daml.ledger.participant.state.kvutils.api.{
+  KeyValueParticipantStateReader,
+  LedgerReader,
+  LedgerRecord
+}
+import com.daml.ledger.participant.state.kvutils.export.WriteSet
+import com.daml.ledger.participant.state.v1.{LedgerId, LedgerInitialConditions, Offset, Update}
+import com.daml.metrics.Metrics
+
+import scala.collection.mutable.ListBuffer
+
+final class LogAppendingReadServiceFactory(
+    metrics: Metrics,
+) extends ReplayingReadServiceFactory {
+  private val recordedBlocks = ListBuffer.empty[LedgerRecord]
+
+  override def appendBlock(writeSet: WriteSet): Unit = {
+    this.synchronized {
+      writeSet.foreach {
+        case (key, value) =>
+          val offset = OffsetBuilder.fromLong(recordedBlocks.length.toLong)
+          recordedBlocks.append(LedgerRecord(offset, key, value))
+      }
+    }
+  }
+
+  /** Returns a new ReadService that can stream all previously recorded updates */
+  override def getReadService(implicit materializer: Materializer): ReplayingReadService =
+    this.synchronized {
```

Comment:
This `synchronized` block does not actually cover calls like `recordedBlocks.toList`. Is this intentional?

Reply:
I think it is not and it seems possible to me that `recordedBlocks` gets accessed when is not yet stable. I'll snapshot it before building future computations on it.

Reply:
Done.

Reply:
Nice catch! Ty.

Reply:
Can you narrow the scope of this `synchronized` block, please?